README.md
# Proof of Liabilities protocol implemented in Rust

[![Crates.io](https://img.shields.io/crates/v/dapol?style=flat-square)](https://crates.io/crates/dapol)
[![License](https://img.shields.io/badge/license-MIT-blue?style=flat-square)](LICENSE)
[![Build Status](https://img.shields.io/github/actions/workflow/status/silversixpence-crypto/dapol/ci.yml?branch=main&style=flat-square)](https://github.com/silversixpence-crypto/dapol/actions/workflows/ci.yml?query=branch%3Amain)

Licensed under [MIT](LICENSE).

## About

Implementation of the DAPOL+ protocol introduced in the "Generalized Proof of Liabilities" by Yan Ji and Konstantinos Chalkias ACM CCS 2021 paper, available [here](https://eprint.iacr.org/2021/1350)

See the [top-level doc for the project](https://hackmd.io/p0dy3R0RS5qpm3sX-_zreA) if you would like to know more about Proof of Liabilities.

## Still to be done

This project is currently still a work in progress, but is ready for
use as is. The code has _not_ been audited yet (as of Nov 2023) and so it is not recommended to use it in production. Progress can be tracked [here](https://github.com/silversixpence-crypto/dapol/issues/91).

Important tasks still to be done:
- Write a spec: https://github.com/silversixpence-crypto/dapol/issues/17
- Support the Deterministic mapping SMT accumulator type: https://github.com/silversixpence-crypto/dapol/issues/9
- Fuzz some of the unit tests: https://github.com/silversixpence-crypto/dapol/issues/46
- Sort out version issues with dependencies: https://github.com/silversixpence-crypto/dapol/issues/11

## How this code can be used

There is both a Rust API and a CLI. Details for both can be found in the sections below.

### Rust API

The API has the following capabilities:
- build a tree using the builder pattern or a configuration file
- generate inclusion proofs from a list of entity IDs (tree required)
- verify an inclusion proof using a root hash (no tree required)

See the [examples](https://github.com/silversixpence-crypto/dapol/examples) directory or [docs](https://docs.rs/dapol/latest/dapol/#rust-api) for details on how to use the API.

### CLI

There is no downloadable executable ([yet](https://github.com/silversixpence-crypto/dapol/issues/110)) so the CLI has to be built from source. You will need to have the rust compiler installed:
```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs >> ./rustup-init.sh
./rustup-init.sh -y --no-modify-path
rm -f ./rustup-init.sh
```

For now you must clone the repo to use the CLI. Once you have cloned the repo build everything:
```bash
# run inside the repo
cargo build --release
```

You can invoke the CLI like so:
```bash
./target/release/dapol help
```

The CLI offers 3 main operations: tree building, proof generation & proof verification. All options can be explored with:
```bash
./target/release/dapol build-tree help
./target/release/dapol gen-proofs help
./target/release/dapol verify-proof help
```

#### Tree building

Building a tree can be done:
- from a config file (see tree_config_example.toml)
- from CLI arguments
- by deserializing an already-built tree

Build a tree using config file (full log verbosity):
```bash
./target/release/dapol -vvv build-tree config-file ./examples/tree_config_example.toml
```

Add serialization:
```bash
./target/release/dapol -vvv build-tree config-file ./examples/tree_config_example.toml --serialize .
```

Deserialize a tree from a file:
```bash
./target/release/dapol -vvv build-tree deserialize <file>
```

Generate proofs (proofs will live in the `./inclusion_proofs/` directory):
```bash
./target/release/dapol -vvv build-tree config-file ./examples/tree_config_example.toml --gen-proofs ./examples/entities_example.csv
```

Build a tree using cli args as apposed to a config file:
```bash
# this will generate random secrets & 1000 random entities
./target/release/dapol -vvv build-tree new --accumulator ndm-smt --height 16 --random-entities 1000
```

#### Proof generation

As seen above, the proof generation can be done via the tree build command, but it can also be done via its own command, which offers some more options around how the proofs are generated.

```bash
./target/release/dapol -vvv gen-proofs --entity-ids ./examples/entities_example.csv --tree-file <serialized_tree_file>
```

The proof generation command only offers 1 way to inject the tree (deserialization), as apposed to the tree build which offers different options.

#### Proof verification

```bash
./target/release/dapol -vvv verify-proof --file-path <inclusion_proof_file> --root-hash <hash>
```

The root hash is logged out at info level when the tree is built or deserialized.

## Benchmarks

To run the benchmarks first clone the repo and then run:
```bash
# Run the benchmarks written in the Criterion framework.
cargo bench --bench criterion_benches

# Run the benchmarks written without a framework.
cargo bench --bench manual_benches

# available env vars (with their default values):
MIN_TOTAL_THREAD_COUNT=0
MIN_ENTITIES=0
MAX_ENTITIES=250000000
MIN_HEIGHT=2
MAX_HEIGHT=64
LOG_VERBOSITY=none # supports error, warn, info, debug
```

The benches are split into 2 parts: Criterion (for small benches) and manual (for large benches). Some of the values of $n$ cause the benchmarks to take *really* long (up to an hour), and so using Criterion (which takes a minimum of 10 samples per bench) makes things too slow. It is advised to run Criterion benches for $n<1000000$ and manual benches otherwise.

A set of tuples is used as input to the benches:

![](resources/readme_eq_benchmark.svg)

You may experience an error building the benches if you are on a fresh Linux machine. If the jemalloc-sys package fails to build then maybe [this](https://github.com/tikv/jemallocator/issues/29) will help.

## Unit test fuzzing

Follow the steps in the [Rust Fuzz Book](https://rust-fuzz.github.io/book/introduction.html) to get started. Essentially:
```bash
# The cargo-fuzz / libfuzzer duo is used
cargo install cargo-fuzz

# Need nightly for cargo-fuzz
rustup default nightly

# Run the max_nodes_to_store block, and don't do more than 300k runs.
cargo fuzz run max_nodes_to_store -- -runs=300000
```




CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## v0.2.0 (2023-12-27)

- Add max_thread_count API parameter [62be10c](https://github.com/silversixpence-crypto/dapol/commit/62be10c9393b2b7e2a4feeedde53fd8a793cbf30)
- Add benchmarks [8ff0379](https://github.com/silversixpence-crypto/dapol/commit/8ff037967fa536fca1122373d72e3e4acb8f169c)
- Fix bug with thread pool not being refilled [5c79aa8](https://github.com/silversixpence-crypto/dapol/commit/5c79aa86cae9b24654b8fa869a010c2edb4815bf)
- Add fuzzing tests [5c79aa8](https://github.com/silversixpence-crypto/dapol/commit/5c79aa86cae9b24654b8fa869a010c2edb4815bf)
- Add benchmark data [bb8c26d](https://github.com/silversixpence-crypto/dapol/commit/bb8c26d3fde82392334be9b39a1ab862f073f854)
- Make code adhere to Rust standards [5a16b36](https://github.com/silversixpence-crypto/dapol/commit/5a16b364771455cb5db0aa8c0ce24e0469d49521)


## v0.1.0 (2023-11-16)

Initial code publish [14fe157](https://github.com/silversixpence-crypto/dapol/commit/14fe1572430992ed0c2bc0c360dc3695f6362004)


src/lib.rs
// Copyright ⓒ 2023 SilverSixpence
// Licensed under the MIT license
// (see LICENSE or <http://opensource.org/licenses/MIT>) All files in the project carrying such
// notice may not be copied, modified, or distributed except according to those terms.

//! # Proof of Liabilities protocol implemented in Rust
//!
//! Implementation of the DAPOL+ protocol introduced in the "Generalized Proof of Liabilities" by Yan Ji and Konstantinos Chalkias ACM CCS 2021 paper, available [here](https://eprint.iacr.org/2021/1350)
//!
//! See the [top-level doc for the project](https://hackmd.io/p0dy3R0RS5qpm3sX-_zreA) if you would like to know more about Proof of Liabilities.
//!
//! ## What is contained in this code
//!
//! This library offers an efficient build algorithm for constructing a binary Merkle Sum Tree representing the liabilities of an organization. Efficiency is achieved through parallelization. Details on the algorithm used can be found in [the multi-threaded builder file](https://github.com/silversixpence-crypto/dapol/blob/main/src/binary_tree/tree_builder/multi_threaded.rs).
//!
//! The paper describes a few different accumulator variants. The Sparse Merkle Sum Tree is the DAPOL+ accumulator, but there are a few different axes of variation, such as how the list of entities is embedded within the tree. The 4 accumulator variants are simply slightly different versions of the Sparse Merkle Sum Tree. Only the Non-Deterministic Mapping Sparse Merkle Tree variant has been implemented so far.
//!
//! The code offers inclusion proof generation & verification using the Bulletproofs protocol for the range proofs.
//!
//! ## Still to be done
//!
//! This project is currently still a work in progress, but is ready for
//! use as is. The code has _not_ been audited yet (as of Nov 2023). Progress can be tracked [here](https://github.com/silversixpence-crypto/dapol/issues/91).
//!
//! A Rust crate has not been released yet, progress can be tracked [here](https://github.com/silversixpence-crypto/dapol/issues/13).
//!
//! A spec for this code still needs to be [written](https://github.com/silversixpence-crypto/dapol/issues/17).
//!
//! A fuzzing technique should be used for the unit [tests](https://github.com/silversixpence-crypto/dapol/issues/46).
//!
//! Performance can be [improved](https://github.com/silversixpence-crypto/dapol/issues/44).
//!
//! Alternate accumulators mentioned in the paper should be built:
//! - [Deterministic mapping SMT](https://github.com/silversixpence-crypto/dapol/issues/9)
//! - [ORAM-based SMT](https://github.com/silversixpence-crypto/dapol/issues/8)
//! - [Hierarchical SMTs](https://github.com/silversixpence-crypto/dapol/issues/7)
//!
//! Other than the above there are a few minor tasks to do, each of which has an issue for tracking.
//!
//! ## How this code can be used
//!
//! There is both a Rust API and a CLI. Details for the API can be found below, and details for the CLI can be found [here](https://github.com/silversixpence-crypto/dapol#cli).
//!
//! ### Rust API
//!
//! The library has not been released as a crate yet (as of Nov 2023) but the API has the following capabilities:
//! - build a tree using the builder pattern or a configuration file
//! - generate inclusion proofs from a list of entity IDs (tree required)
//! - verify an inclusion proof using a root hash (no tree required)
//!
//! ```
//! use std::str::FromStr;
//! use std::path::Path;
//!
//! use dapol::utils::LogOnErrUnwrap;
//!
//! fn main() {
//!     let log_level = clap_verbosity_flag::LevelFilter::Debug;
//!     dapol::utils::activate_logging(log_level);
//!
//!     // =========================================================================
//!     // Tree building.
//!
//!     let ndm_smt = build_ndm_smt_using_builder_pattern();
//!     let accumulator = build_accumulator_using_config_file();
//!
//!     // The above 2 builder methods produce a different tree because the entities
//!     // are mapped randomly to points on the bottom layer, but the entity mapping
//!     // of one tree should simply be a permutation of the other. We check this:
//!     let ndm_smt_other = match accumulator {
//!         dapol::Accumulator::NdmSmt(ndm_smt_other) => {
//!             assert_ne!(ndm_smt_other.root_hash(), ndm_smt.root_hash());
//!
//!             for (entity, _) in ndm_smt_other.entity_mapping() {
//!                 assert!(ndm_smt.entity_mapping().contains_key(&entity));
//!             }
//!
//!             ndm_smt_other
//!         }
//!     };
//!
//!     // =========================================================================
//!     // Inclusion proof generation & verification.
//!
//!     let entity_id = dapol::EntityId::from_str("john.doe@example.com").unwrap();
//!     simple_inclusion_proof_generation_and_verification(&ndm_smt, entity_id.clone());
//!     advanced_inclusion_proof_generation_and_verification(&ndm_smt_other, entity_id);
//! }
//!
//! /// Example on how to use the builder pattern to construct an NDM-SMT tree.
//! pub fn build_ndm_smt_using_builder_pattern() -> dapol::accumulators::NdmSmt {
//!     let src_dir = env!("CARGO_MANIFEST_DIR");
//!     let resources_dir = Path::new(&src_dir).join("examples");
//!
//!     let secrets_file = resources_dir.join("ndm_smt_secrets_example.toml");
//!     let entities_file = resources_dir.join("entities_example.csv");
//!
//!     let height = dapol::Height::expect_from(16);
//!
//!     let config = dapol::accumulators::NdmSmtConfigBuilder::default()
//!         .height(height)
//!         .secrets_file_path(secrets_file)
//!         .entities_path(entities_file)
//!         .build();
//!
//!     config.parse().unwrap()
//! }
//!
//! /// An inclusion proof can be generated from only a tree + entity ID.
//! pub fn simple_inclusion_proof_generation_and_verification(
//!     ndm_smt: &dapol::accumulators::NdmSmt,
//!     entity_id: dapol::EntityId,
//! ) {
//!     let inclusion_proof = ndm_smt.generate_inclusion_proof(&entity_id).unwrap();
//!     inclusion_proof.verify(ndm_smt.root_hash()).unwrap();
//! }
//!
//! /// The inclusion proof generation algorithm can be customized via some
//! /// parameters. See [dapol][InclusionProof] for more details.
//! pub fn advanced_inclusion_proof_generation_and_verification(
//!     ndm_smt: &dapol::accumulators::NdmSmt,
//!     entity_id: dapol::EntityId,
//! ) {
//!     // Determines how many of the range proofs in the inclusion proof are
//!     // aggregated together. The ones that are not aggregated are proved
//!     // individually. The more that are aggregated the faster the proving
//!     // and verification times.
//!     let aggregation_percentage = dapol::percentage::ONE_HUNDRED_PERCENT;
//!     let aggregation_factor = dapol::AggregationFactor::Percent(aggregation_percentage);
//!     let aggregation_factor = dapol::AggregationFactor::default();
//!
//!     // 2^upper_bound_bit_length is the upper bound used in the range proof i.e.
//!     // the secret value is shown to reside in the range [0, 2^upper_bound_bit_length].
//!     let upper_bound_bit_length = 32u8;
//!     let upper_bound_bit_length = dapol::DEFAULT_RANGE_PROOF_UPPER_BOUND_BIT_LENGTH;
//!
//!     let inclusion_proof = ndm_smt
//!         .generate_inclusion_proof_with(&entity_id, aggregation_factor, upper_bound_bit_length)
//!         .unwrap();
//!
//!     inclusion_proof.verify(ndm_smt.root_hash()).unwrap();
//! }
//!
//! /// Example on how to build a tree using a config file.
//! ///
//! /// The config file can be used for any accumulator type since the type is
//! /// specified by the config file.
//! ///
//! /// This is also an example usage of [dapol][utils][LogOnErrUnwrap].
//! pub fn build_accumulator_using_config_file() -> dapol::Accumulator {
//!     let src_dir = env!("CARGO_MANIFEST_DIR");
//!     let resources_dir = Path::new(&src_dir).join("examples");
//!     let config_file = resources_dir.join("tree_config_example.toml");
//!
//!     dapol::AccumulatorConfig::deserialize(config_file)
//!         .log_on_err_unwrap()
//!         .parse()
//!         .log_on_err_unwrap()
//! }
//! ```

mod kdf;
mod node_content;

pub mod cli;
pub mod percentage;
pub mod read_write_utils;
pub mod utils;

mod hasher;
pub use hasher::Hasher;

mod max_thread_count;
pub use max_thread_count::{initialize_machine_parallelism, MaxThreadCount, MACHINE_PARALLELISM};

pub mod accumulators;
pub use accumulators::{Accumulator, AccumulatorConfig, AccumulatorConfigError, AccumulatorError};

mod binary_tree;
pub use binary_tree::{Height, HeightError, MAX_HEIGHT, MIN_HEIGHT};

mod secret;
pub use secret::{Secret, SecretParserError};

mod inclusion_proof;
pub use inclusion_proof::{
    AggregationFactor, InclusionProof, InclusionProofError,
    DEFAULT_RANGE_PROOF_UPPER_BOUND_BIT_LENGTH,
};

mod entity;
pub use entity::{Entity, EntityId, EntityIdsParser, EntityIdsParserError};

/// Used for surfacing fuzzing tests to the fuzzing module in the ./fuzz
/// directory.
#[cfg(fuzzing)]
pub mod fuzz {
    pub use super::binary_tree::multi_threaded::tests::fuzz_max_nodes_to_store;
}
